### ISSUE TITLE
net/http/fcgi: race detected during execution of TestResponseWriterSniffsContentType test

### ISSUE THREAD

**Issue by dmitshur** [golang/go#41167]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.15.1 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes.

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/dmitshur/Library/Caches/go-build"
GOENV="/Users/dmitshur/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/dmitshur/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/dmitshur/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/zb/5p8cwfhj29gf_m8vdy8ylmlr00jwcj/T/go-build487751365=/tmp/go-build -gno-record-gcc-switches -fno-common"
</pre></details>

### What did you do?

I looked at https://build.golang.org/?branch=release-branch.go1.15 (and https://build.golang.org/?branch=release-branch.go1.14) and spotted a few new failures on race builders. Then I reproduced it locally by running:

```
$ go test -count=10 -run=TestResponseWriterSniffsContentType -race net/http/fcgi
```

### What did you expect to see?

```
ok  	net/http/fcgi	0.171s
```

### What did you see instead?

```
[...]
--- FAIL: TestResponseWriterSniffsContentType (0.00s)
    --- FAIL: TestResponseWriterSniffsContentType/jpg (0.00s)
    testing.go:1023: race detected during execution of test
FAIL
FAIL	net/http/fcgi	0.155s
```

/cc @FiloSottile @empijei

**Comment by tz70s** [comment]
Probably a connection lock is needed for this read, https://github.com/golang/go/blob/master/src/net/http/fcgi/child.go#L158

writes from `typeStdin` are asynchronous. https://github.com/golang/go/blob/master/src/net/http/fcgi/child.go#L229

**Comment by empijei** [comment]
Hi, I have looked into this a bit more and I tried to backport the racy test to 1.15.

The race was already there and was not introduced by the security release.

The only difference is that now we cover that codepath with tests, so I think this race has been there for a long time.

**Comment by dmitshur** [comment]
Sounds good. Since the race isn't new and critical, the fix can be made available in Go 1.16 and we don't need to backport it. However, let's skip the new test so that the race builders continue to pass and provide full coverage on the release branches.

@gopherbot Please backport to both Go 1.15 and 1.14, but only the test fix.
